src/c_api/c_api.cc

```diff
@@ -1877,20 +1879,22 @@ int MXNDArraySave(const char* fname,
                   NDArrayHandle* args,
                   const char** keys) {
   API_BEGIN();
-  std::vector<NDArray> data(num_args);
-  std::vector<std::string> names;
-  for (uint32_t i = 0; i < num_args; ++i) {
-    data[i] = *static_cast<NDArray*>(args[i]);
-  }
-  if (keys != nullptr) {
-    names.resize(num_args);
-    for (uint32_t i = 0; i < num_args; ++i) {
-      names[i] = keys[i];
-    }
-  }
-  {
-    std::unique_ptr<dmlc::Stream> fo(dmlc::Stream::Create(fname, "w"));
-    mxnet::NDArray::Save(fo.get(), data, names);
+
+  CHECK_NOTNULL(fname);
+
+  if (num_args == 1 && static_cast<NDArray *>(args[0])->storage_type() == kDefaultStorage) {
+      NDArray *array = static_cast<NDArray *>(args[0]);
+      npy::save_array(fname, *array);
+  } else {
+      int write_mode = ZIP_TRUNCATE | ZIP_CREATE;
+      for (uint32_t i = 0; i < num_args; ++i) {
+          NDArray *array = static_cast<NDArray *>(args[i]);
+          const std::string array_key = keys == nullptr ? std::to_string(i) : keys[i];
+          npz::save_array(write_mode, fname, array_key, *array);
```

Comment:
does it handle non-default storage type?

Reply:
yes